spec.html

```diff
@@ -76,3 +80,25 @@ <h1>String.prototype.replaceAll ( _searchValue_, _replaceValue_ )</h1>
     1. Return _result_.
   </emu-alg>
 </emu-clause>
+
+<emu-clause id="sec-string.prototype.matchall">
+  <h1>String.prototype.matchAll ( _regexp_ )</h1>
+  <p>Performs a regular expression match of the String representing the *this* value against _regexp_ and returns an iterator. Each iteration result's value is an Array object containing the results of the match, or *null* if the String did not match.</p>
+  <p>When the `matchAll` method is called, the following steps are taken:</p>
+  <emu-alg>
+    1. Let _O_ be ? RequireObjectCoercible(*this* value).
+    1. If _regexp_ is neither *undefined* nor *null*, then
+      1. <ins>Let _isRegExp_ be ? IsRegExp(_regexp_).</ins>
+      1. <ins>If _isRegExp_ is true, then</ins>
+        1. <ins>Let _flags_ be ? ToString(? Get(_regexp_, `"flags"`)).</ins>
+        1. <ins>If _flags_ does not contain `"g"`, throw a *TypeError* exception.</ins>
+      1. Let _matcher_ be ? GetMethod(_regexp_, @@matchAll).
```

Comment:
it does seem kind of odd that a non RegExp can have any kind of non “all” behavior it wants, but an actual RegExp is forced to have the proper “all” behavior. Not sure anything can be done about that, but it is an inconsistency introduced by this change.

Reply:
I see what you're saying. Still, there's value in making built-ins work well / in unsurprising ways by default, even if for userland subclasses, this is up to the user.

Reply:
If we only/mostly care about built-ins, I wonder if makes more sense to move this check into <https://tc39.es/ecma262/#sec-regexp-prototype-matchall>?

Reply:
Interesting. I don't have a strong preference tbh. Any opinions? @schuay @ljharb

Reply:
No strong opinion. One argument for the current spot is that the location is consistent between `matchAll` and `replaceAll`. There's no `@@replaceAll` so we could not consistently move both checks to a RegExp builtin. I'm fine with either way.

Reply:
It seems like the motivation/value of moving it inside the symbol methods is so a subclass could *override* the behavior - ie, could lack the g flag but still be “all” (vice versa is already going to be possible). I don’t see any benefit in allowing regex subclasses to deviate from this rather strong guard that we’ll have decided on - without concrete use cases, i think here is probably a better spot.